### ISSUE TITLE
Feature: Ability to restart, reopen or reset a queue.

### ISSUE THREAD

**Issue by TimZaman** [tensorflow/tensorflow#4535]
Taking into account what @mrry said on http://stackoverflow.com/questions/39204335/can-a-tensorflow-queue-be-reopened-after-it-is-closed/39209186#39209186 ;

It's not possible to reset a queue unless you have multiple sessions. How about resetting it when you have a single session?

**Use case:**
In a single session, if one would want to run _exactly_ one epoch (where epoch_size%batch_size!=0) of a model _x_ times and report it's result exactly after one full and single epoch. It would be convenient if one could restart the queues.

**Comment by mrry** [comment]
@suharshs is working on many different improvements to the Session code, and I think making `Session::Reset()` sensible for a single session is on his list.

**Comment by TimZaman** [comment]
This is perfect! 0.11? Thanks for your valuable contributions on tf all over the interwebs, @mrry .

**Comment by sjperkins** [comment]
Has this implemented for single session in 0.11.0? The documentation implies that the queue reset mechanism suggested on [SO](http://stackoverflow.com/questions/39204335/can-a-tensorflow-queue-be-reopened-after-it-is-closed) (reset queues in a resource container, needs distributed session) is still the way to do it.

**Comment by TimZaman** [comment]
@suharshs Any updates on this? Cheers.

**Comment by suharshs** [comment]
Apologies, I haven't gotten a chance to look into this yet.

**Comment by kmhofmann** [comment]
I'm _really_ looking forward to an implementation of in-session queue resetting, and hope this will be coming soon.

This would enable repeated evaluation of an enqueued validation set, where (epoch_size % batch_size > 0) in a single session. I thought this was a rather common case, and am a bit surprised there is no good support in TensorFlow for this scenario.

Currently we are resorting to complicated hacks that [emit undefined behavior](http://stackoverflow.com/questions/41187745/tensorflow-how-can-i-evaluate-a-validation-data-queue-multiple-times-during-tra), and I cannot get the proposed workaround using multiple sessions to work.

**Comment by shaayaansayed** [comment]
+1, this would be incredibly useful. 

Does anyone have the multiple session workaround going by any chance?

**Comment by bxshi** [comment]
Any updates on this? Thanks!

**Comment by suharshs** [comment]
@mrry is working on some input pipeline stuff that should make this obsolete. So assigning to him. For now , continue to use the workaround provided here: http://stackoverflow.com/questions/39204335/can-a-tensorflow-queue-be-reopened-after-it-is-closed

**Comment by mrry** [comment]
Thanks @suharshs! We're planning to move away from queues and provide first-class support for multi-epoch processing in the redesigned input pipeline API. As a result, I doubt that we will add these features to the queues directly, but instead we'll make it easy to get a signal between epochs in the new API.

I'm going to close this issue for now. Please feel free to comment on #7951 if there are particular features that you'd particularly like to see or other use cases that you'd like us to support in the new API!

**Comment by albertz** [comment]
@mrry That sounds interesting. Can I already try it? Or when is this planned to be released?

**Comment by mrry** [comment]
@albertz Good timing: we just released the initial version as `tf.contrib.data` in the 1.2rc0 release.

Here are some docs to get started: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/data/README.md

**Comment by albertz** [comment]
That's interesting. I also extended my [StackOverflow answer here](https://stackoverflow.com/questions/41187745/tensorflow-how-can-i-evaluate-a-validation-data-queue-multiple-times-during-tra/44067467#44067467) to address this.

Small questions:

When I use `repeat` (for multiple epochs) together with `shuffle` (as `read_batch_features` does internally), how will I notice when some epochs ends, and what the current epoch is? Also, when the epoch ends, will the `ShuffleDataset` wait first to dequeue everything or will it already be filled with more data from the next epoch? In the last epoch, or if I don't use `repeat`, will the `ShuffleDataset` dequeue all remaining data, like `tf.RandomShuffleQueue` dequeueing does after close?

If I want to have more control, I would not use `repeat` but go once over the data and use `ShuffleDataset` to get shuffling like `RandomShuffleQueue`, and then at some point I get `OutOfRangeError` and I know that I reached the end of the epoch. Then I reinitializable the iterator, like it is described.

If I use `PaddedBatchDataset`, how could I recover the lengths of each sequence in the batch?

The way how to get data doesn't really fit any way how I get the data usually. In my case, I have a thread and I receive data there and I don't know in advance when it will end but I see when it ends. Then I wait until I processed all the buffers and then I have finished one epoch. How can I get this logic with the `Dataset`?

How can I wrap around a `Dataset` over a queue? I have some thread with reads some data from somewhere and which can feed it and queue it somehow. How do I get the data into the `Dataset`? I could repeat some dummy tensor infinite times and then use `map` to just return my `queue.dequeue()` but that really only gets me back to all the original problems with the queue, i.e. how to reopen the queue.

**Comment by ppwwyyxx** [comment]
I agree with @albertz . Queues have one important feature -- it takes data from Python side and that's important for a lot of use cases where data cannot be easily preprocessed with TF operators. Also it's much easier to write for anyone.
Therefore dataset API at least today doesn't seem to be strong enough to replace Queues. I still hope to see some methods to close & reopen the queue.

**Comment by TimZaman** [comment]
Since when do queues take data from the Python side?

**Comment by ppwwyyxx** [comment]
Oh at least it's possible to `feed_dict` into a queue. I update the comment a bit.

**Comment by TimZaman** [comment]
Hmm - technically, you cannot do that. You can feed_dict into a placeholder. Then _that_ can go into a queue. This holds up for anything, so also for the Dataset API:
https://www.tensorflow.org/versions/r1.3/programmers_guide/datasets
The dataset API can be _feedable_, just like a queue.

**Comment by ppwwyyxx** [comment]
You're right -- almost anything is feedable. I was only thinking about the typical cases I work with. I'll correct my point:
Queues have one feature -- prefetch. And that seemed to be missing from dataset API. Therefore dataset API cannot replace queues now.
Or to summarize, what I want is some reader that can: 
1. take data from Python 
2. prefetch 
3. Let me know I finished 1 pass of the dataset and start another (it can be reset for dataset API, or close-reopen for queues, whatever)
1 is actually not a problem for either one.

**Comment by TimZaman** [comment]
(1) and (2) seem independent points to me. That's probably not correct. Look for capacity* or buffer* through the docs: https://www.tensorflow.org/api_docs/python/tf/contrib/data/Dataset.

**Comment by ppwwyyxx** [comment]
They seem independent to me as well.
All "capacity*" or "buffer*" I found is about shufflequeue. That's not for prefetch.

**Comment by samwhitlock** [comment]
@ppwwyyxx as far as I can tell it's not feedable. I constructed [this example](https://gist.github.com/samwhitlock/bfafc793c6b7589d12514e741c06eea1) and [modified it](https://gist.github.com/samwhitlock/251aaa04a1c64a929f20657b53662e2b) to try rerunning the initializer with a different feed value and it caused an error. However, the `tf.Errors.OutOfRangeError` was thrown on all `sess.run()` calls after the first.

This seems like a pretty common use case to support, where one input is fed but produces multiple outputs per input.

Is there a way to do this with the existing `Dataset` API that I'm not seeing?

**Comment by mrry** [comment]
Hi folks! You might be interested in my [answer](https://stackoverflow.com/a/45928467/3574081) to @albertz's question on Stack Overflow. In short, there have been a couple of developments since the 1.3 release that might be useful to people on this thread, and I'd encourage you to try them out:

* [`Dataset.from_generator()`](https://www.tensorflow.org/versions/master/api_docs/python/tf/contrib/data/Dataset#from_generator) makes it easy to create a `Dataset` from a Python generator. This gives you a way to wrap "feeding" code by writing a Python function that uses `yield` (rather than `feed_dict`) to pass NumPy arrays into the dataset.
* [`Dataset.prefetch()`](https://www.tensorflow.org/versions/master/api_docs/python/tf/contrib/data/Dataset#prefetch) makes it easy to create a background thread that preprocesses elements. (You could do this in 1.3, using `Dataset.map(lambda x, y, z: (x, y, z), num_threads=1, output_buffer_size=N)`, but the new version is simpler and a little more efficient.)

As an aside, we've upgraded `Dataset.map(..., num_threads=N)` for `N > 1` to use a more efficient representation that doesn't actually create a huge number of threads. (The new argument is called `num_parallel_calls`, but we've kept `num_threads` for now.) This should make it more efficient to process large numbers of elements in parallel, and we've seen encouraging improvements on our benchmarks for CNN image input pipelines.
